oci/layer/tar_extract.go

```diff
@@ -707,3 +710,48 @@ out:
 	}
 	return nil
 }
+
+// patchOverlayWhiteout converts a whiteout char dev into a userxattr to handle
+// the case where there is a whiteout char dev and a subdir is being created
+// with that whiteout filename in the directory path which fails
+func patchOverlayWhiteout(dir string, fsEval fseval.FsEval) error {
+	for p := dir; p != "/"; p = filepath.Dir(p) {
+		pinfo, err := os.Lstat(p)
+		if err != nil {
+			if os.IsNotExist(errors.Cause(err)) {
+				return nil
+			}
+
+			log.Errorf("failed to stat %s", p)
+			return err
+		}
+		isCharDev, err := isOverlayWhiteoutCharDev(pinfo, p, fsEval)
+		if err != nil {
+			return err
+		}
+
+		if !isCharDev {
+			continue
+		}
+
+		log.Infof("patching overlay whiteout %s from char dev to userxattr\n", p)
+		// remove the char device
+		if err := os.Remove(p); err != nil {
+			return err
+		}
+		// replace it with a dir
+		err = fsEval.MkdirAll(p, 0777)
+		if err != nil {
+			return err
+		}
+		// and set userxattr
+		err = fsEval.Lsetxattr(p, "user.overlay.opaque", []byte("y"), 0)
```

Comment:
does umoci need to handle both user.overlay.opaque and trusted.overlay.opaque, in case we are running as root and not on an overlay that has userxattr set? I know stacker will mount with userxattr, but umoci in general might not be able to assume that